core/src/main/java/org/apache/iceberg/view/ViewMetadata.java

```diff
@@ -0,0 +1,139 @@
+import org.apache.iceberg.util.PropertyUtil;
+import org.immutables.value.Value;
+
+@Value.Immutable
+public abstract class ViewMetadata implements Serializable {
+  static final int SUPPORTED_VIEW_FORMAT_VERSION = 1;
+
+  public abstract int formatVersion();
+
+  public abstract String location();
+
+  public abstract Integer currentSchemaId();
+
+  public abstract List<Schema> schemas();
+
+  public abstract int currentVersionId();
+
+  public abstract List<ViewVersion> versions();
+
+  public abstract List<ViewHistoryEntry> history();
+
+  public abstract Map<String, String> properties();
+
+  public ViewVersion version(int versionId) {
+    return versionsById().get(versionId);
+  }
+
+  @Value.Lazy
+  public ViewVersion currentVersion() {
+    return versionsById().get(currentVersionId());
+  }
+
+  @Value.Derived
+  public Map<Integer, ViewVersion> versionsById() {
+    return indexVersions(versions());
+  }
+
+  @Value.Derived
+  public Map<Integer, Schema> schemasById() {
+    return indexSchemas(schemas());
+  }
+
+  @Value.Lazy
+  public Schema schema() {
+    return schemasById().get(currentSchemaId());
+  }
+
+  @Value.Default
+  int versionHistorySizeToKeep() {
+    return PropertyUtil.propertyAsInt(
+        properties(),
+        ViewProperties.VERSION_HISTORY_SIZE,
+        ViewProperties.VERSION_HISTORY_SIZE_DEFAULT);
+  }
+
+  @Value.Check
+  protected ViewMetadata checkAndNormalize() {
+    Preconditions.checkArgument(
+        formatVersion() > 0 && formatVersion() <= ViewMetadata.SUPPORTED_VIEW_FORMAT_VERSION,
+        "Unsupported format version: %s",
+        formatVersion());
+
+    Preconditions.checkArgument(versions().size() > 0, "Invalid view versions: empty");
+    Preconditions.checkArgument(history().size() > 0, "Invalid view history: empty");
+    Preconditions.checkArgument(schemas().size() > 0, "Invalid schemas: empty");
+
+    Preconditions.checkArgument(
+        versionsById().containsKey(currentVersionId()),
+        "Cannot find current version %s in view versions: %s",
+        currentVersionId(),
+        versionsById().keySet());
+
+    Preconditions.checkArgument(
+        schemasById().containsKey(currentSchemaId()),
+        "Cannot find current schema with id %s in schemas: %s",
+        currentSchemaId(),
+        schemasById().keySet());
+
+    Preconditions.checkArgument(
+        versionHistorySizeToKeep() >= 1,
+        "%s must be positive but was %s",
+        ViewProperties.VERSION_HISTORY_SIZE,
+        versionHistorySizeToKeep());
+
+    if (versions().size() > versionHistorySizeToKeep()) {
+      List<ViewVersion> versions =
+          versions().subList(versions().size() - versionHistorySizeToKeep(), versions().size());
+      List<ViewHistoryEntry> history =
+          history().subList(history().size() - versionHistorySizeToKeep(), history().size());
+      return ImmutableViewMetadata.builder().from(this).versions(versions).history(history).build();
```

Comment:
I think this is confusing. The builder should enforce the maximum number of versions, not the check method. It's awkward that this returns a copy of the view metadata.

Reply:
this seems to be the "official" way of normalizing an object: http://immutables.github.io/immutable.html#normalization

Reply:
I agree here completely. 
It would be great if there could be a better & shorter alternative to achieve this. 
The (longer) alternative would be to have our own builder that internally uses `ImmutableViewMetadata.builder()` and does the "normalization" but that doesn't prevent anyone from using `ImmutableViewMetadata.builder()` directly